File: windows/src-tauri/src/fortschritt.rs
//! Was eine Seite ueber ihren Stand verraet — ausgewertet und als Datei abgelegt.
//! Nachbau von Fortschritt.swift; die Dateien sind mit denen vom Mac austauschbar.

use crate::bibliothek::{ganzzahl, json_lesen, Seite};
use crate::orte::{atomar_schreiben, json_schoen, orte, vorbereiten};
use crate::zeit;
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use std::collections::{BTreeMap, HashMap};
use std::fs;

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
pub struct Tagespensum {
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub datum: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub ziel: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub geschafft: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub treffer: Option<i64>,
}

impl Tagespensum {
    pub fn erledigt(&self) -> bool {
        match (self.ziel, self.geschafft) {
            (Some(z), Some(g)) if z > 0 => g >= z && self.datum.as_deref() == Some(&zeit::heute()),
            _ => false,
        }
    }
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SeitenStand {
    pub seite: String,
    #[serde(default)]
    pub titel: String,
    #[serde(default)]
    pub fach: String,
    #[serde(default)]
    pub thema: String,
    #[serde(default)]
    pub typ: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub gesamt: Option<i64>,
    #[serde(default)]
    pub sitzen: i64,
    #[serde(default)]
    pub wackel: i64,
    #[serde(default)]
    pub geuebt: i64,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub tagespensum: Option<Tagespensum>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub zuletzt_geoeffnet: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub zuletzt_geuebt: Option<String>,
    #[serde(default)]
    pub wackel_items: Vec<String>,
    #[serde(default)]
    pub roh: BTreeMap<String, String>,
}

impl SeitenStand {
    pub fn anteil_sitzen(&self) -> f64 {
        match self.gesamt {
            Some(g) if g > 0 => self.sitzen as f64 / g as f64,
            _ => 0.0,
        }
    }
}

/// Wertet den rohen localStorage einer Seite aus.
/// `roh` enthaelt alle Schluessel, die mit "lern:" anfangen.
pub fn auswerten(roh: &HashMap<String, String>, seite: &Seite) -> SeitenStand {
    let praefix = format!("lern:{}@", seite.id);
    let mut stand = SeitenStand {
        seite: seite.id.clone(),
        titel: seite.titel.clone(),
        fach: seite.fach.clone(),
        thema: seite.thema.clone(),
        typ: seite.typ.clone(),
        // Alle Seiten teilen sich einen localStorage — nur behalten, was zu
        // dieser Seite gehoert.
        roh: roh
            .iter()
            .filter(|(k, _)| k.starts_with(&praefix))
            .map(|(k, v)| (k.clone(), v.clone()))
            .collect(),
        zuletzt_geoeffnet: Some(zeit::jetzt_iso()),
        ..Default::default()
    };

    let hauptschluessel = format!("lern:{}@v{}", seite.id, seite.version);
    let Some(objekt) = roh
        .get(&hauptschluessel)
        .and_then(|t| serde_json::from_str::<Value>(t).ok())
        .and_then(|v| v.as_object().cloned())
    else {
        // Seite erfuellt den Vertrag nicht — wir wissen nur, dass sie offen war.
        return stand;
    };

    stand.gesamt = objekt.get("gesamt").and_then(ganzzahl);

    if let Some(items) = objekt.get("items").and_then(Value::as_object) {
        for (name, wert) in items {
            let Some(wert) = wert.as_object() else { continue };
            let sass = wert.get("sass").and_then(ganzzahl).unwrap_or(0);
            let sass_nicht = wert.get("sassNicht").and_then(ganzzahl).unwrap_or(0);
            if sass + sass_nicht <= 0 {
                continue;
            }
            stand.geuebt += 1;
            let wacklig = match wert.get("letzter").and_then(Value::as_str) {
                Some(letzter) => letzter != "sass",
                None => sass_nicht >= sass,
            };
            if wacklig {
                stand.wackel += 1;
                stand.wackel_items.push(name.clone());
            } else {
                stand.sitzen += 1;
            }
        }
        if stand.gesamt.is_none() {
            stand.gesamt = Some(items.len() as i64);
        }
    }
    stand.wackel_items.sort();

    if let Some(p) = objekt.get("tagespensum").and_then(Value::as_object) {
        stand.tagespensum = Some(Tagespensum {
            datum: p.get("datum").and_then(Value::as_str).map(String::from),
            ziel: p.get("ziel").and_then(ganzzahl),
            geschafft: p.get("geschafft").and_then(ganzzahl),
            treffer: p.get("treffer").and_then(ganzzahl),
        });
    }
    stand.zuletzt_geuebt = objekt
        .get("zuletztGeoeffnet")
        .and_then(Value::as_str)
        .map(String::from);
    stand
}

/// Legt den Stand einer Seite als eigene Datei ab.
pub fn sichern(stand: &SeitenStand) {
    vorbereiten();
    let name = format!("{}.json", stand.seite.replace('/', "__"));
    let Ok(wert) = serde_json::to_value(stand) else { return };
    let _ = atomar_schreiben(&orte().fortschritt.join(name), &json_schoen(&wert));
    uebersicht_schreiben();
}

/// Alle Staende — fuer Seitenleiste, Startseite und fuer Claude.
pub fn alle_staende() -> HashMap<String, SeitenStand> {
    let mut ergebnis = HashMap::new();
    let Ok(eintraege) = fs::read_dir(&orte().fortschritt) else { return ergebnis };
    for e in eintraege.flatten() {
        let pfad = e.path();
        let name = e.file_name().to_string_lossy().into_owned();
        if !name.ends_with(".json") || name == "uebersicht.json" {
            continue;
        }
        let Some(stand) = json_lesen(&pfad).and_then(|v| serde_json::from_value::<SeitenStand>(v).ok())
        else {
            continue;
        };
        ergebnis.insert(stand.seite.clone(), stand);
    }
    ergebnis
}

/// Eine einzelne Datei, die Claude zu Sessionbeginn liest.
fn uebersicht_schreiben() {
    let mut staende: Vec<SeitenStand> = alle_staende().into_values().collect();
    staende.sort_by(|a, b| {
        b.zuletzt_geoeffnet
            .as_deref()
            .unwrap_or("")
            .cmp(a.zuletzt_geoeffnet.as_deref().unwrap_or(""))
    });
    let mut u = Map::new();
    u.insert("erstellt".into(), Value::String(zeit::jetzt_iso()));
    u.insert(
        "staende".into(),
        serde_json::to_value(staende).unwrap_or(Value::Array(vec![])),
    );
    let _ = atomar_schreiben(
        &orte().fortschritt.join("uebersicht.json"),
        &json_schoen(&Value::Object(u)),
    );
}

// MARK: - Zustand der App (Favoriten, zuletzt geoeffnet, Hell/Dunkel)

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Zustand {
    #[serde(default)]
    pub favoriten: Vec<String>,
    /// Seiten-IDs, neueste zuerst.
    #[serde(default)]
    pub zuletzt: Vec<String>,
    #[serde(default = "dunkel")]
    pub theme: String,
    #[serde(skip_serializing_if = "Option::is_none", default)]
    pub letzte_seite: Option<String>,
}

fn dunkel() -> String {
    "dark".into()
}

impl Default for Zustand {
    fn default() -> Self {
        Zustand { favoriten: vec![], zuletzt: vec![], theme: dunkel(), letzte_seite: None }
    }
}

impl Zustand {
    pub fn laden() -> Zustand {
        json_lesen(&orte().zustand)
            .and_then(|v| serde_json::from_value(v).ok())
            .unwrap_or_default()
    }

    pub fn sichern(&self) {
        vorbereiten();
        if let Ok(wert) = serde_json::to_value(self) {
            let _ = atomar_schreiben(&orte().zustand, &json_schoen(&wert));
        }
    }

    pub fn besucht(&mut self, id: &str) {
        self.zuletzt.retain(|z| z != id);
        self.zuletzt.insert(0, id.to_string());
        self.zuletzt.truncate(8);
        self.letzte_seite = Some(id.to_string());
        self.sichern();
    }

    pub fn favorit_umschalten(&mut self, id: &str) {
        if self.favoriten.iter().any(|f| f == id) {
            self.favoriten.retain(|f| f != id);
        } else {
            self.favoriten.push(id.to_string());
        }
        self.sichern();
    }
}

File: Quelle/Fortschritt.swift
import Foundation

// MARK: - Was eine Seite über ihren Stand verrät

struct Tagespensum: Codable {
    var datum: String?
    var ziel: Int?
    var geschafft: Int?
    var treffer: Int?

    var erledigt: Bool {
        guard let ziel, let geschafft, ziel > 0 else { return false }
        return geschafft >= ziel && datum == Datum.heute
    }
}

struct SeitenStand: Codable {
    var seite: String
    var titel: String
    var fach: String
    var thema: String
    var typ: String

    /// Anzahl abfragbarer Items laut Seite.
    var gesamt: Int?
    /// Items, deren letzter Versuch "saß" war.
    var sitzen: Int = 0
    /// Items, deren letzter Versuch "saß nicht" war — die Wackelkandidaten.
    var wackel: Int = 0
    /// Items, die überhaupt schon einmal drankamen.
    var geuebt: Int = 0

    var tagespensum: Tagespensum?
    var zuletztGeoeffnet: String?
    var zuletztGeuebt: String?

    /// Namen der Wackelkandidaten — damit Claude gezielt nachfragen kann.
    var wackelItems: [String] = []

    /// Alles, was die Seite gespeichert hat, unverändert. Rettungsanker für
    /// Seiten, die den Vertrag (noch) nicht erfüllen.
    var roh: [String: String] = [:]

    var anteilSitzen: Double {
        guard let gesamt, gesamt > 0 else { return 0 }
        return Double(sitzen) / Double(gesamt)
    }
}

// MARK: - Auswerten und Ablegen

enum Fortschritt {

    /// Wertet den rohen localStorage einer Seite aus.
    /// `roh` enthält alle Schlüssel, die mit "lern:" anfangen.
    static func auswerten(roh: [String: String], seite: Seite) -> SeitenStand {
        var stand = SeitenStand(seite: seite.id, titel: seite.titel, fach: seite.fach,
                                thema: seite.thema, typ: seite.typ)
        // Alle Seiten teilen sich denselben Ursprung (127.0.0.1) und damit einen
        // localStorage. Hier nur behalten, was wirklich zu dieser Seite gehoert —
        // sonst steht in jeder Datei der Stand aller anderen Seiten mit drin.
        stand.roh = roh.filter { $0.key.hasPrefix("lern:\(seite.id)@") }
        stand.zuletztGeoeffnet = ISO.jetzt

        let hauptschluessel = "lern:\(seite.id)@v\(seite.version)"
        guard let text = roh[hauptschluessel],
              let daten = text.data(using: .utf8),
              let objekt = try? JSONSerialization.jsonObject(with: daten) as? [String: Any]
        else {
            // Seite erfüllt den Vertrag nicht — wir wissen nur, dass sie offen war.
            return stand
        }

        stand.gesamt = objekt["gesamt"] as? Int

        if let items = objekt["items"] as? [String: [String: Any]] {
            for (name, wert) in items {
                let sass      = wert["sass"] as? Int ?? 0
                let sassNicht = wert["sassNicht"] as? Int ?? 0
                guard sass + sassNicht > 0 else { continue }
                stand.geuebt += 1
                // "zuletzt" entscheidet, ob das Item aktuell sitzt.
                if let letzter = wert["letzter"] as? String {
                    if letzter == "sass" { stand.sitzen += 1 }
                    else { stand.wackel += 1; stand.wackelItems.append(name) }
                } else if sassNicht >= sass {
                    stand.wackel += 1; stand.wackelItems.append(name)
                } else {
                    stand.sitzen += 1
                }
            }
            if stand.gesamt == nil { stand.gesamt = items.count }
        }
        stand.wackelItems.sort()

        if let p = objekt["tagespensum"] as? [String: Any] {
            stand.tagespensum = Tagespensum(datum: p["datum"] as? String,
                                            ziel: p["ziel"] as? Int,
                                            geschafft: p["geschafft"] as? Int,
                                            treffer: p["treffer"] as? Int)
        }
        stand.zuletztGeuebt = objekt["zuletztGeoeffnet"] as? String
        return stand
    }

    /// Legt den Stand einer Seite als eigene Datei ab.
    static func sichern(_ stand: SeitenStand) {
        Orte.vorbereiten()
        let name = stand.seite.replacingOccurrences(of: "/", with: "__") + ".json"
        let ziel = Orte.fortschritt.appendingPathComponent(name)
        let kodierer = JSONEncoder()
        kodierer.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let daten = try? kodierer.encode(stand) else { return }
        try? daten.write(to: ziel, options: .atomic)
        uebersichtSchreiben()
    }

    /// Alle Stände geladen — für Sidebar, Startseite und für Claude.
    static func alleStaende() -> [String: SeitenStand] {
        let dateien = (try? FileManager.default.contentsOfDirectory(
            at: Orte.fortschritt, includingPropertiesForKeys: nil)) ?? []
        var ergebnis: [String: SeitenStand] = [:]
        for datei in dateien where datei.pathExtension == "json"
                                && datei.lastPathComponent != "uebersicht.json" {
            guard let daten = try? Data(contentsOf: datei),
                  let stand = try? JSONDecoder().decode(SeitenStand.self, from: daten)
            else { continue }
            ergebnis[stand.seite] = stand
        }
        return ergebnis
    }

    /// Eine einzelne Datei, die Claude zu Sessionbeginn liest.
    private static func uebersichtSchreiben() {
        struct Uebersicht: Codable {
            var erstellt: String
            var staende: [SeitenStand]
        }
        let u = Uebersicht(erstellt: ISO.jetzt,
                           staende: alleStaende().values
                            .sorted { ($0.zuletztGeoeffnet ?? "") > ($1.zuletztGeoeffnet ?? "") })
        let kodierer = JSONEncoder()
        kodierer.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let daten = try? kodierer.encode(u) else { return }
        try? daten.write(to: Orte.fortschritt.appendingPathComponent("uebersicht.json"),
                         options: .atomic)
    }
}

enum ISO {
    static let formatierer: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withFullDate, .withTime, .withColonSeparatorInTime,
                           .withDashSeparatorInDate]
        f.timeZone = TimeZone.current
        return f
    }()
    static var jetzt: String { formatierer.string(from: Date()) }
    static func datum(_ text: String?) -> Date? {
        guard let text else { return nil }
        return formatierer.date(from: text)
    }
}

// MARK: - Zustand der App (Favoriten, zuletzt geöffnet)

struct Zustand: Codable {
    var favoriten: [String] = []
    var zuletzt: [String] = []          // Seiten-IDs, neueste zuerst
    var theme: String = "dark"
    var letzteSeite: String?
    var neuigkeitGesehen: String?       // Kennung des zuletzt angesehenen Updates (ⓘ)
    var anleitungAusgeblendet: Bool?    // Kasten „Erste Schritte“ per × weggeklickt

    static var aktuell: Zustand = laden()

    static func laden() -> Zustand {
        guard let daten = try? Data(contentsOf: Orte.zustandDatei),
              let z = try? JSONDecoder().decode(Zustand.self, from: daten)
        else { return Zustand() }
        return z
    }

    func sichern() {
        Orte.vorbereiten()
        let kodierer = JSONEncoder()
        kodierer.outputFormatting = [.prettyPrinted, .sortedKeys]
        guard let daten = try? kodierer.encode(self) else { return }
        try? daten.write(to: Orte.zustandDatei, options: .atomic)
    }

    mutating func besucht(_ seitenID: String) {
        zuletzt.removeAll { $0 == seitenID }
        zuletzt.insert(seitenID, at: 0)
        if zuletzt.count > 8 { zuletzt = Array(zuletzt.prefix(8)) }
        letzteSeite = seitenID
        sichern()
    }

    mutating func favoritUmschalten(_ seitenID: String) {
        if favoriten.contains(seitenID) { favoriten.removeAll { $0 == seitenID } }
        else { favoriten.append(seitenID) }
        sichern()
    }

    func istFavorit(_ seitenID: String) -> Bool { favoriten.contains(seitenID) }
}
